ledger/artifactmanager/artifactmanager_senders.go

```diff
@@ -0,0 +1,156 @@
+/*
+ *    Copyright 2019 Insolar Technologies
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package artifactmanager
+
+import (
+	"context"
+	"sync"
+
+	"github.com/insolar/insolar/core"
+	"github.com/insolar/insolar/core/message"
+	"github.com/insolar/insolar/core/reply"
+	"github.com/insolar/insolar/instrumentation/inslogger"
+	"github.com/insolar/insolar/ledger/storage"
+	"github.com/pkg/errors"
+	"go.opencensus.io/stats"
+)
+
+type ledgerArtifactSenders struct {
+	pulseStorage core.PulseStorage
+	jetStorage   storage.JetStorage
+	defaultBus   core.MessageBus
+
+	codeCacheLock sync.Mutex
+	codeCache     map[core.RecordRef]*cacheEntry
+}
+
+func newLedgerArtifactSenders(
+	pulseStorage core.PulseStorage,
+	jetStorage storage.JetStorage,
+	defaultBus core.MessageBus,
+) *ledgerArtifactSenders {
+	return &ledgerArtifactSenders{
+		pulseStorage: pulseStorage,
+		jetStorage:   jetStorage,
+		defaultBus:   defaultBus,
+		codeCache:    map[core.RecordRef]*cacheEntry{},
+	}
+}
+
+func (m *ledgerArtifactSenders) bus(ctx context.Context) core.MessageBus {
+	return core.MessageBusFromContext(ctx, m.defaultBus)
+}
+
+func (m *ledgerArtifactSenders) cachedSender(sender Sender) Sender {
+	return func(ctx context.Context, msg core.Message, options *core.MessageSendOptions) (core.Reply, error) {
+		codeMsg := msg.(*message.GetCode)
+
+		m.codeCacheLock.Lock()
+		entry, ok := m.codeCache[codeMsg.Code]
+		if !ok {
+			entry = &cacheEntry{}
+			m.codeCache[codeMsg.Code] = entry
+		}
+		m.codeCacheLock.Unlock()
+
+		entry.Lock()
+		defer entry.Unlock()
+
+		if entry.code != nil {
+			return entry.code, nil
+		}
+
+		response, err := sender(ctx, msg, options)
+		if err != nil {
+			return nil, err
+		}
+		castedResp, ok := response.(*reply.Code)
+		if !ok {
+			return response, err
+		}
+
+		entry.code = castedResp
+		return response, err
+	}
+}
+
+func (m *ledgerArtifactSenders) followRedirectSender(sender Sender) Sender {
```

Comment:
If can be a function with a closure on MessageBus, passed as an argument (see instrumentHandler).